proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,322 @@
+# Explicit Backing Fields
+
+- **Type**: Design Proposal
+- **Author**: ?
+- **Contributors**: Svetlana Isakova, Kirill Rakhman, Dmitry Petrov, Roman Elizarov, Ben Leggiero, Matej Drobnič, Mikhail Glukhikh, Nikolay Lunyak
+- **Status**: Implemented in FIR
+- **Prototype**: Implemented
+- **Initial YouTrack Issue**: [KT-14663](https://youtrack.jetbrains.com/issue/KT-14663)
+- **Initial Proposal**: [private_public_property_types#122](https://github.com/Kotlin/KEEP/pull/122)
+
+## Summary
+
+**Note**: initial proposal contents have been partially copied down here for convenience. Despite the different approach, the already shown use cases are still relevant.
+
+>  Common pattern in java is having full type accessible as private property and then only exposing required interface in the public getter:
+>
+> ```java
+> class MyClass {
+> 	// Use full type for private access
+> 	private final ArrayList<String> data = new ArrayList<>();
+> 	
+> 	// Only expose what is needed in public getter
+> 	public List<String> getData() {
+> 		return data;
+> 	}
+> }
+> ```
+>
+> This pattern allows easy hiding of implementation details and allows for easy clean external interfaces to the classes. But there is no idiomatic Kotlin way to achieve this pattern. Best thing you can do is define two separate properties to mimic how Java does it:
+>
+> ```kotlin
+> class MyClass {
+>     private val _data = ArrayList<String>()
+>     val data: List<String>
+>         get() = _data
+> }
+> ```
+
+With the proposed syntax in mind, the above code snippet could be rewritten as follows:
+
+```kotlin
+class MyClass {
+    val data: List<String>
+        field = ArrayList()
+}
+```
+
+## Use Cases
+
+### Motivation
+
+> - There is no clean idiomatic way to do this pattern in Kotlin
+> - Current best approach throws away all benefits of Kotlin properties and forces developer to write Java-like code with separate private field and public getter
+> - Current best approach forces developer to assign two different names to single property (or pad private property, for example adding `_` prefix and then using this prefix everywhere in code)
+> - This is another place where Java pattern could be introduced into Kotlin with less boilerplate
+
+### Read-Only from Outside
+
+> We often do not want our data structures to be modified from outside. Unlike Java, this can be easily achieved in Kotlin by just exposing read-only `List`. But as already ilustrated in above example, exposing different type of a property is a bit messy.
+
+```kotlin
+private val _items = mutableListOf<Item>()
+val item : List<Item> by _items
+```
+
+And the new syntax allows us to write:
+
+```kotlin
+val items: List<Item>
+	private field = mutableListOf()
```

Comment:
What's `private` is supposed to mean here before `field`? Also, from here and down the document something is wrong with the coding style (indentation suddenly becomes 8 spaces instead of 4).

Reply:
It's true that `private` here is redundant. I thought, specifying it explicitly here would make the example look more consistent with the sample code above. It's like "look, we've had a `private` backing property, and now here is a `private` backing field"

Reply:
Changed `private` to `internal` in this example